src/rdkafka_fetcher.c

```diff
@@ -171,55 +171,88 @@ static void rd_kafka_fetch_reply_handle_partition_error(
+                         * (with backoff).
+                         */
+                        rd_rkb_dbg(rkb, MSG, "FETCH",
+                                   "Topic %s [%" PRId32
+                                   "]: %s not "
+                                   "available on broker %" PRId32
+                                   " (leader %" PRId32
+                                   "): updating metadata and retrying",
+                                   rktp->rktp_rkt->rkt_topic->str,
+                                   rktp->rktp_partition,
+                                   rd_kafka_fetch_pos2str(
+                                       rktp->rktp_offsets.fetch_pos),
+                                   rktp->rktp_broker_id, rktp->rktp_leader_id);
+                }
+
+                if (err == RD_KAFKA_RESP_ERR_UNKNOWN_LEADER_EPOCH) {
+                        rd_rkb_dbg(rkb, MSG | RD_KAFKA_DBG_CONSUMER, "FETCH",
+                                   "Topic %s [%" PRId32
+                                   "]: Fetch failed at %s: %s: broker %" PRId32
+                                   "has not yet caught up on latest metadata: "
+                                   "retrying",
+                                   rktp->rktp_rkt->rkt_topic->str,
+                                   rktp->rktp_partition,
+                                   rd_kafka_fetch_pos2str(
+                                       rktp->rktp_offsets.fetch_pos),
+                                   rd_kafka_err2str(err), rktp->rktp_broker_id);
+                }
+
+                if (rktp->rktp_broker_id != rktp->rktp_leader_id) {
+                        rd_kafka_toppar_delegate_to_leader(rktp);
+                }
                 /* Request metadata information update*/
                 rd_kafka_toppar_leader_unavailable(rktp, "fetch", err);
                 break;
 
-        case RD_KAFKA_RESP_ERR_OFFSET_NOT_AVAILABLE:
-                /* Occurs when:
-                 *   - Msg exists on broker but
-                 *     offset > HWM, or:
-                 *   - HWM is >= offset, but msg not
-                 *     yet available at that offset
-                 *     (replica is out of sync).
-                 *
-                 * Handle by retrying FETCH (with backoff).
-                 */
-                rd_rkb_dbg(rkb, MSG, "FETCH",
-                           "Topic %s [%" PRId32 "]: Offset %" PRId64
-                           " not "
-                           "available on broker %" PRId32 " (leader %" PRId32
-                           "): retrying",
-                           rktp->rktp_rkt->rkt_topic->str, rktp->rktp_partition,
-                           rktp->rktp_offsets.fetch_offset,
-                           rktp->rktp_broker_id, rktp->rktp_leader_id);
-                break;
-
         case RD_KAFKA_RESP_ERR_OFFSET_OUT_OF_RANGE: {
-                int64_t err_offset;
+                rd_kafka_fetch_pos_t err_pos;
```

Comment:
I think we are partly handling case4, see `rd_kafka_fetch_preferred_replica_handle`. However, rather than backing off, we should resort to the reset policy rather than doing the truncation/validation again with the leader, as per the document: ("If we do not have an epoch corresponding to the current fetch offset, then we will skip the truncation check and use the same handling that we used for the FETCH_OFFSET_TOO_SMALL case (basically we resort to the reset policy). In practice, out of range errors due to ISR propagation delays should be extremely rare because it requires a consumer to see a committed offset before a follower that is still considered in the ISR.")

But I'm not so sure about case3, I think we're not handling it at all.

Reply:
Tests 0104 is checking those cases and in fetch from follower it's different from unclean leader election because the epoch doesn't change and the validation is not called, but the out of range is retried on the leader.